File: src/domain/validate.ts
import type { Dish, Menu, Product, ValidationReport } from "../types/models";

export function validateWorkspace(products: Product[], dishes: Dish[], menus: Menu[]): ValidationReport {
  const errors: string[] = [];
  const warnings: string[] = [];
  const productIds = new Set<string>();
  const dishIds = new Set<string>();
  const menuIds = new Set<string>();

  for (const product of products) {
    if (productIds.has(product.id)) errors.push(`дублируется продукт ${product.id}`);
    productIds.add(product.id);
    if (!product.name.trim()) errors.push(`у продукта ${product.id} пустое имя`);
    if (product.pricePerKg === 0) warnings.push(`у «${product.name}» нет цены`);
    if (product.nutritionPer100g.kcal === 0 && product.category !== "spices") {
      warnings.push(`у «${product.name}» не заполнена пищевая ценность`);
    }
  }

  for (const dish of dishes) {
    if (dishIds.has(dish.id)) errors.push(`дублируется блюдо ${dish.id}`);
    dishIds.add(dish.id);
    if (!dish.name.trim()) errors.push(`у блюда ${dish.id} пустое имя`);
    if (dish.lines.length === 0) errors.push(`блюдо «${dish.name}» без продуктов`);
    if (dish.basePortions <= 0) errors.push(`у «${dish.name}» некорректные порции`);
    for (const line of dish.lines) {
      if (!productIds.has(line.productId)) {
        errors.push(`блюдо «${dish.name}»: продукт ${line.productId} не найден`);
      }
    }
    if (dish.targetPortionG === 0) warnings.push(`у «${dish.name}» нет планового выхода порции`);
  }

  for (const menu of menus) {
    if (menuIds.has(menu.id)) errors.push(`дублируется меню ${menu.id}`);
    menuIds.add(menu.id);
    if (menu.guests <= 0) errors.push(`у меню «${menu.name}» некорректное число гостей`);
    for (const day of menu.days) {
      for (const slot of day.slots) {
        if (!dishIds.has(slot.dishId)) {
          errors.push(`меню «${menu.name}», ${day.dateLabel}: блюдо ${slot.dishId} не найдено`);
        }
      }
    }
  }

  return { ok: errors.length === 0, errors, warnings };
}

File: rust/src/validate.rs
use crate::dish::DishBook;
use crate::error::{Error, Result};
use crate::menu::MenuBook;
use crate::product::ProductCatalog;

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ValidationReport {
    pub ok: bool,
    pub errors: Vec<String>,
    pub warnings: Vec<String>,
}

impl ValidationReport {
    pub fn new() -> Self {
        Self {
            ok: true,
            errors: Vec::new(),
            warnings: Vec::new(),
        }
    }

    pub fn error(&mut self, message: impl Into<String>) {
        self.ok = false;
        self.errors.push(message.into());
    }

    pub fn warn(&mut self, message: impl Into<String>) {
        self.warnings.push(message.into());
    }
}

impl Default for ValidationReport {
    fn default() -> Self {
        Self::new()
    }
}

pub fn validate_workspace(
    catalog: &ProductCatalog,
    dishes: &DishBook,
    menus: &MenuBook,
) -> ValidationReport {
    let mut report = ValidationReport::new();

    let mut ids = std::collections::HashSet::new();
    for product in &catalog.products {
        if !ids.insert(product.id.clone()) {
            report.error(format!("дублируется продукт {}", product.id));
        }
        if let Err(err) = product.validate() {
            report.error(err.to_string());
        }
        if product.price_per_kg == 0.0 {
            report.warn(format!("у «{}» нет цены — калькуляция будет нулевой", product.name));
        }
        if product.nutrition_per_100g.kcal == 0.0
            && !matches!(
                product.category,
                crate::product::ProductCategory::Spices | crate::product::ProductCategory::Drinks
            )
        {
            report.warn(format!("у «{}» не заполнена пищевая ценность", product.name));
        }
    }

    let mut dish_ids = std::collections::HashSet::new();
    for dish in &dishes.dishes {
        if !dish_ids.insert(dish.id.clone()) {
            report.error(format!("дублируется блюдо {}", dish.id));
        }
        if let Err(err) = dish.validate() {
            report.error(err.to_string());
        }
        for line in &dish.lines {
            if catalog.get(&line.product_id).is_err() {
                report.error(format!(
                    "блюдо «{}»: продукт {} не найден",
                    dish.name, line.product_id
                ));
            }
        }
        if dish.target_portion_g == 0.0 {
            report.warn(format!(
                "у «{}» нет планового выхода порции",
                dish.name
            ));
        }
    }

    let mut menu_ids = std::collections::HashSet::new();
    for menu in &menus.menus {
        if !menu_ids.insert(menu.id.clone()) {
            report.error(format!("дублируется меню {}", menu.id));
        }
        if let Err(err) = menu.validate() {
            report.error(err.to_string());
        }
        for day in &menu.days {
            for slot in &day.slots {
                if dishes.get(&slot.dish_id).is_err() {
                    report.error(format!(
                        "меню «{}», день {}: блюдо {} не найдено",
                        menu.name, day.date_label, slot.dish_id
                    ));
                }
            }
        }
    }

    report
}

pub fn require_ok(report: &ValidationReport) -> Result<()> {
    if report.ok {
        Ok(())
    } else {
        Err(Error::Validation(report.errors.join("; ")))
    }
}
